File: src/main/java/com/example/order/service/OrderServiceImpl.java
package com.example.order.service;

import com.example.annotation.MainDiscountPolicy;
import com.example.discount.DiscountPolicy;
import com.example.member.domain.Member;
import com.example.member.repository.MemberRepository;
import com.example.order.domain.Order;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.stereotype.Service;

@Service
public class OrderServiceImpl implements OrderService, InitializingBean, DisposableBean {
	
	private final MemberRepository memberRepository;
	private final DiscountPolicy discountPolicy;

	public OrderServiceImpl(MemberRepository memberRepository, @MainDiscountPolicy DiscountPolicy discountPolicy) {
		this.memberRepository = memberRepository;
		this.discountPolicy = discountPolicy;
	}

	@Override
	public Order createOrder(Long memberId, String itemName, int itemPrice) {
		Member member = memberRepository.findById(memberId).orElseThrow();
		int discountPrice = discountPolicy.discount(member, itemPrice);
		
		return new Order(memberId, itemName, itemPrice, discountPrice);
	}

	@Override
	public void afterPropertiesSet() throws Exception {
		System.out.println("OrderServiceImpl.afterPropertiesSet");
	}
	
	@Override
	public void destroy() throws Exception {
		System.out.println("OrderServiceImpl.destroy");
	}

}

File: 김영한/spring_core_principle_basic/src/main/kotlin/hello/core/order/OrderServiceImpl.kt
package hello.core.order

import hello.core.discount.DiscountPolicy
import hello.core.member.repository.MemberRepository
import org.springframework.stereotype.Component

@Component
class OrderServiceImpl(
    private val memberRepository: MemberRepository,
    private val discountPolicy: DiscountPolicy
): OrderService {

    override fun createOrder(memberId: Long, itemName: String, itemPrice: Int): Order {
        val member = memberRepository.findById(memberId)
        val discountPrice = discountPolicy.discount(member, itemPrice)

        return Order(memberId, itemName, itemPrice, discountPrice)
    }

}